src/Components/Web.JS/src/Rendering/DomMerging/DomSync.ts

```diff
@@ -0,0 +1,224 @@
+    const endAtNodeExclOrNull = destination instanceof Element ? null : destination.endExclusive;
+    while (nextDestinationNode !== endAtNodeExclOrNull) {
+      treatAsMatch(nextDestinationNode!, nextNewContentNode!);
+      nextDestinationNode = nextDestinationNode!.nextSibling;
+      nextNewContentNode = nextNewContentNode!.nextSibling;
+    }
+    if (nextNewContentNode) {
+      // Should never be possible, as it would imply a bug in the edit script calculation, or possibly an unsupported
+      // scenario like a DOM mutation observer modifying the destination nodes while we are working on them
+      throw new Error('Updating the DOM failed because the sets of trailing nodes had inconsistent lengths.');
+    }
+  }
+}
+
+function treatAsMatch(destination: Node, source: Node) {
+  switch (destination.nodeType) {
+    case Node.TEXT_NODE:
+    case Node.COMMENT_NODE:
+      break;
+    case Node.ELEMENT_NODE:
+      synchronizeAttributes(destination as Element, source as Element);
+      synchronizeDomContent(destination as Element, source as Element);
+      break;
+    default:
+      throw new Error(`Not implemented: matching nodes of type ${destination.nodeType}`);
+  }
+}
+
+function treatAsSubstitution(destination: Node, source: Node) {
+  switch (destination.nodeType) {
+    case Node.TEXT_NODE:
+    case Node.COMMENT_NODE:
+      (destination as Text).textContent = (source as Text).textContent;
+      break;
+    default:
+      throw new Error(`Not implemented: substituting nodes of type ${destination.nodeType}`);
+  }
+}
+
+function domNodeComparer(a: Node, b: Node): UpdateCost {
+  if (a.nodeType !== b.nodeType) {
+    return UpdateCost.Infinite;
+  }
+
+  switch (a.nodeType) {
+    case Node.TEXT_NODE:
+    case Node.COMMENT_NODE:
+      // We're willing to update text and comment nodes in place, but treat the update operation as being
+      // as costly as an insertion or deletion
+      return a.textContent === b.textContent ? UpdateCost.None : UpdateCost.Some;
+    case Node.ELEMENT_NODE:
+      // For elements, we're only doing a shallow comparison and don't know if attributes/descendants are different.
+      // We never 'update' one element type into another. We regard the update cost for same-type elements as zero because
+      // then the 'find common prefix/suffix' optimization can include elements in those prefixes/suffixes.
+      // TODO: If we want to support some way to force matching/nonmatching based on @key, we can add logic here
+      //       to return UpdateCost.Infinite if either has a key but they don't match. This will prevent unwanted retention.
+      //       For the converse (forcing retention, even if that means reordering), we could post-process the list of
+      //       inserts/deletes to find matches based on key to treat those pairs as 'move' operations.
+      return (a as Element).tagName === (b as Element).tagName ? UpdateCost.None : UpdateCost.Infinite;
```

Comment:
Maybe this question has already been considered before (or maybe a solution is already in the works), so I apologize if there was a conversation I missed!

I could see how comparing the element type is enough to eliminate ambiguity in many cases, but could there also be issues where user-entered state "shifts" when content gets synchronized? For example:

```html
<!-- Original content -->
<input type="text" placeholder="Email">   <!-- 'me@myself.com' -->
<input type="text" placeholder="Phone #"> <!-- '12345678' -->

<!-- Modified content -->
<input type="text" placeholder="Email">   <!-- 'me@myself.com' -->
<input type="number" placeholder="Zip">   <!-- '12345678' -->
<input type="text" placeholder="Phone #"> <!-- '' -->
```

In this example, the user-entered phone number becomes the zip code. If we considered the input's "type" attribute as part of the comparison, I think it would resolve this particular example, but the issue would persist if the inserted element was also a "text" input.

I see your comment here about `@key` - would we consider going further and including sequence numbers as special attributes (at least for input-like HTML elements) and use those in the comparison here?

Reply:
That's a really great point. I hadn't considered the unique aspects of `input`-type elements and how it could go wrong.

Rather than trying to guess the developer's intent, it would be much safer simply to say we treat `input` like any other element, in that the new content always takes precedence and replaces all old state. So what I plan to do is:

 * Update the "update attributes" logic so that it has the same "set special values" logic as we use in `BrowserRenderer.ts` (i.e., when assigning a `value` or `checked`, depending on the element type, we also write it as a property not just an attribute)
 * Update the test to show that we do still preserve user edits if no attribute change occurred, but we don't preserve it if the value attribute did change

In theory the issue you describe could still occur, but it would be much more of an edge case as it would only occur if the old and new elements had identical `value` attributes but somehow represented unrelated parts of the UI.

Reply:
OK, following this feedback, I've taken an even more conservative approach:

 * Uses the `tryApplySpecialProperty` logic that was previously exclusive to `BrowserRenderer.ts`. That is, it now respects all the special cases where we need to assign input-type element values as properties rather than attributes. This ensures that whenever we update an "attribute", it does overwrite any user edits, making the DOM really synchronized.
 * When deciding whether an "attribute" is changed, implemented the inverse of the above, so we respect special cases to read back the property values rather than attribute values for the same set of input-type elements. This avoids even the "more of an edge case" scenario I mentioned above, since now we no longer care whether the value attribute has changed - for input-type elements we work in terms of the property instead.

I think this now covers just about everything, although the sheer sophistication of the update logic now makes me strongly suspect we'll eventually uncover even more special cases we will want to handle. What makes me happy enough to proceed anyway is:

 1. The set of special cases accounted for comes from the existing battle-tested `BrowserRenderer.ts`, which we know people have been fine with in production for years
 2. We've still got some preview releases to go
 3. In the worst case, people can still disable this DOM node retention behavior

One thing in particular that I know won't be ideal is the handling of custom elements that maintain internal state. For those, we're just going to update the attributes when preserving those elements, since we have no way of telling them to reset their internal state or what that would even mean for them. However, the same is true for our interactive rendering approach, so we're not really causing new problems here (hopefully).